interpreter/cling/lib/Interpreter/DefinitionShadowing.cpp

```diff
@@ -0,0 +1,145 @@
+// LICENSE.TXT for details.
+//------------------------------------------------------------------------------
+
+#include "DefinitionShadowing.h"
+#include <iostream>
+
+#include "cling/Utils/AST.h"
+
+#include "clang/AST/Decl.h"
+#include "clang/AST/DeclContextInternals.h"
+#include "clang/AST/Stmt.h"
+#include "clang/Sema/Lookup.h"
+#include "clang/Sema/Sema.h"
+
+using namespace clang;
+
+namespace cling {
+  /// \brief Returns whether `L' comes from a Cling input line.
+  static bool typedInClingPrompt(FullSourceLoc L) {
+    if (L.isInvalid())
+      return false;
+    llvm::StringRef S{L.getManager().getPresumedLoc(L).getFilename()};
+    return S.startswith("ROOT_prompt_")
+           || S.startswith("input_line_");
+  }
+
+  /// \brief Returns whether the given {Function,Tag,Var}Decl is a definition.
+  static bool isDefinition(Decl *D) {
+    if (auto _D = dyn_cast<FunctionDecl>(D)) return _D->isThisDeclarationADefinition();
+    if (auto _D = dyn_cast<TagDecl>(D)) return _D->isThisDeclarationADefinition();
+    if (auto _D = dyn_cast<VarDecl>(D)) return _D->isThisDeclarationADefinition();
+    return true;
+  }
+
+  DefinitionShadowing::DefinitionShadowing(Sema* S)
+    : ASTTransformer(S), m_Context(&S->getASTContext()),
+      m_TU(S->getASTContext().getTranslationUnitDecl())
+  {
+    const_cast<PrintingPolicy&>(m_Context->getPrintingPolicy()).SuppressUnwrittenScope = true;
+  }
+
+  /// \brief Make a declaration hidden for SemaLookup; internally used in `invalidatePreviousDefinitions()'.
+  /// This directly manipulates lookup tables to avoid a patch to Clang!
+  void DefinitionShadowing::hideDecl(clang::NamedDecl *D) {
+    if (Scope* S = m_Sema->getScopeForContext(m_TU)) {
+      S->RemoveDecl(D);
+      if (utils::Analyze::isOnScopeChains(D, *m_Sema))
+        m_Sema->IdResolver.RemoveDecl(D);
+    }
+    clang::StoredDeclsList &SDL = (*m_TU->getLookupPtr())[D->getDeclName()];
+    if (SDL.getAsVector() || SDL.getAsDecl() == D)
+      SDL.remove(D);
+  }
+
+  /// \brief Lookup the given name and invalidate all clashing declarations (as seen from the TU).
+  /// \return Returns whether a previous definition exists.
+  bool DefinitionShadowing::invalidatePreviousDefinitions(NamedDecl *D) {
+    LookupResult Previous(*m_Sema, D->getDeclName(), D->getLocation(),
+                   Sema::LookupOrdinaryName, Sema::ForRedeclaration);
+    m_Sema->LookupQualifiedName(Previous, m_TU);
+
+    for (auto I : Previous) {
+      if (I == D) continue;
+      auto NS = dyn_cast<NamespaceDecl>(I->getLexicalDeclContext());
+      if (isDefinition(I)
+          && (!isDefinition(D) || !(NS && NS->getName().startswith("__cling_N5")))) continue;
+
+      hideDecl(I);
+
+      if (TagDecl *TD = dyn_cast<TagDecl>(I)) {
+        if (!TD->isTransparentContext()) continue;
+        for (auto J : TD->decls()) hideDecl(cast<NamedDecl>(J));
+      }
+    }
+    return !Previous.empty();
```

Comment:
Should this handle the `if (I == D) continue;` case?

Reply:
Should be fixed.  :-)